libraries/AP_GPS/AP_GPS_SBP2.cpp

```diff
@@ -0,0 +1,494 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+//
+//  Swift Navigation SBP GPS driver for ArduPilot.
+//	Code by Niels Joubert
+//
+//  Swift Binary Protocol format: http://docs.swift-nav.com/
+//
+
+#include "AP_GPS.h"
+#include "AP_GPS_SBP2.h"
+#include <DataFlash/DataFlash.h>
+
+extern const AP_HAL::HAL& hal;
+
+#define SBP_DEBUGGING 1
+#define SBP_DEBUGGING_SUPERVERBOSE 1
+#define SBP_HW_LOGGING 1
+
+//INVARIANT: We expect SBP to give us a heartbeat in less than 2 seconds.
+//  This is more lax than the default Piksi settings,
+//  and we assume th user hasn't reconfigured their Piksi to longer heartbeat intervals
+#define SBP_TIMEOUT_HEARTBEAT  2000
+
+#if SBP_DEBUGGING
+ # define Debug(fmt, args ...)                  \
+do {                                            \
+    hal.console->printf("%s:%d: " fmt "\n",     \
+                        __FUNCTION__, __LINE__, \
+                        ## args);               \
+    hal.scheduler->delay(1);                    \
+} while(0)
+#else
+ # define Debug(fmt, args ...)
+#endif
+
+AP_GPS_SBP2::AP_GPS_SBP2(AP_GPS &_gps, AP_GPS::GPS_State &_state,
+                       AP_HAL::UARTDriver *_port) :
+    AP_GPS_Backend(_gps, _state, _port),
+
+    waiting_for_first_heartbeat(true),
+    last_injected_data_ms(0),
+    last_full_update_tow(0),
+    crc_error_counter(0)
+{
+
+    Debug("SBP Driver Initialized");
+
+    parser_state.state = sbp_parser_state_t::WAITING;
+
+    //Externally visible state
+    state.status = AP_GPS::NO_FIX;
+    state.last_gps_time_ms = last_heartbeat_received_ms = AP_HAL::millis();
+
+    state.have_vertical_velocity   = true;
+    state.have_speed_accuracy      = false;
+    state.have_horizontal_accuracy = true;
+    state.have_vertical_accuracy   = true;
+
+}
+
+// Process all bytes available from the stream
+//
+bool
+AP_GPS_SBP2::read(void)
+{
+
+    //Invariant: Calling this function processes *all* data current in the UART buffer.
+    //
+    //IMPORTANT NOTICE: This function is NOT CALLED for several seconds
```

Comment:
The comment was in the other driver so it's not a big deal, but this more of an issue to be remarked upon inside the AP_GPS layer, as it's not really upto your driver if you are unloaded or not. Still if it's relevant because you are commanding something from the GPS then feel free to leave it in.